evm-contracts/src/v0.8/dev/ValidatorProxy.sol

```diff
@@ -0,0 +1,195 @@
+  );
+  event AggregatorUpgraded(
+    address indexed previous,
+    address indexed current
+  );
+  event NewValidatorProposed(
+    address indexed validator
+  );
+  event ProposedValidatorRetracted(
+    address indexed validator
+  );
+  event ValidatorUpgraded(
+    address indexed previous,
+    address indexed current
+  );
+  /// @notice The proposed aggregator called validate, but the call was not passed on to any validators
+  event ProposedAggregatorValidateCall(
+    address indexed proposed,
+    uint256 previousRoundId,
+    int256 previousAnswer,
+    uint256 currentRoundId,
+    int256 currentAnswer
+  );
+
+  constructor(
+    address aggregator,
+    address validator
+  )
+    ConfirmedOwner(msg.sender)
+  {
+    s_currentAggregator.target = aggregator;
+    s_currentValidator.target = validator;
+  }
+
+  function validate(
+    uint256 previousRoundId,
+    int256 previousAnswer,
+    uint256 currentRoundId,
+    int256 currentAnswer
+  )
+    external
+    override
+    returns (
+      bool
+    )
+  {
+    address currentAggregator = s_currentAggregator.target;
+    address proposedAggregator = s_proposedAggregator;
+    require(msg.sender == currentAggregator || msg.sender == proposedAggregator, "Not a configured aggregator");
+    // If the aggregator is still in proposed state, emit an event and don't push to any validator.
+    // This is to confirm that `validate` is being called prior to upgrade.
+    if (msg.sender == proposedAggregator) {
+      emit ProposedAggregatorValidateCall(
+        proposedAggregator,
+        previousRoundId,
+        previousAnswer,
+        currentRoundId,
+        currentAnswer
+      );
+      return false;
+    }
+
+    // Send the validate call to the current validator
+    ProxyConfiguration memory currentValidator = s_currentValidator;
+    bool success = AggregatorValidatorInterface(currentValidator.target).validate(
+      previousRoundId,
+      previousAnswer,
+      currentRoundId,
+      currentAnswer
+    );
+    // If there is a new proposed validator, send the validate call to that validator also
+    if (currentValidator.hasNewProposal) {
+      bool proposedSuccess = AggregatorValidatorInterface(s_proposedValidator).validate(
+        previousRoundId,
+        previousAnswer,
+        currentRoundId,
+        currentAnswer
+      );
+      success = success && proposedSuccess;
+    }
+    return success;
+  }
+
+  /** AGGREGATOR CONFIGURATION FUNCTIONS **/
+
+  function proposeNewAggregator(
+    address proposed
+  )
+    external
+    onlyOwner()
+  {
+    s_proposedAggregator = proposed;
+    s_currentAggregator.hasNewProposal = true;
+    emit NewAggregatorProposed(proposed);
+  }
+
+  function retractProposedAggregator()
+    external
+    onlyOwner()
+  {
+    address proposed = s_proposedAggregator;
+    s_proposedAggregator = address(0);
+    s_currentAggregator.hasNewProposal = false;
+    emit ProposedAggregatorRetracted(proposed);
+  }
```

Comment:
might be simpler to only have a single function "proposeNewAggregator" and support cancelling via proposeNewAggregator(0x0)